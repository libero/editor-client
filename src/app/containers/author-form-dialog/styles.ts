import { makeStyles } from '@material-ui/core/styles';

export const useAuthorFormStyles = makeStyles((theme) => ({
  root: {},

  inputField: {
    marginBottom: theme.spacing(3)
  },

  buttonPanel: {
    display: 'flex',
    marginTop: theme.spacing(1),
    '& > *:last-child': {
      marginLeft: theme.spacing(2)
    }
  },

  spacer: {
    flexGrow: 1
  },

  affiliatedAuthorRow: {
    display: 'flex',
    marginTop: theme.spacing(3),
    '&:first-child': {
      marginTop: 0
    }
  },

  affiliatedAuthorInput: {
    flex: 1,
    '& fieldset > legend': {
      maxWidth: 1000
    }
  },

  deleteButton: {
    height: 32,
    width: 32,
    alignSelf: 'center',
    marginLeft: theme.spacing(1)
  }
}));
