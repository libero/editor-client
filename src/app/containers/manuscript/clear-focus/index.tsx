import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import * as manuscriptEditorAction from 'app/actions/manuscript-editor.actions';

export const ClearFocus: React.FC<{ className?: string }> = ({ children, className }) => {
  const dispatch = useDispatch();
  const handleClick = useCallback(() => {
    dispatch(manuscriptEditorAction.removeFocusAction());
  }, [dispatch]);

  return (
    <div className={className} onClick={handleClick}>
      {children}
    </div>
  );
};
