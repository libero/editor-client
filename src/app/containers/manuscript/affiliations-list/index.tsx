import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';

import { SectionContainer } from 'app/components/section-container';
import { ActionButton } from 'app/components/action-button';
import { getAffiliations } from 'app/selectors/manuscript.selectors';
import { Affiliation } from 'app/models/affiliation';
import { useAffiliationStyles } from './styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ConnectedAffiliationFormDialog } from 'app/containers/affiliation-form-dialog';
import { ComponentWithId } from 'app/types/utility.types';

export const AffiliationsList: React.FC<ComponentWithId> = ({ id }) => {
  const classes = useAffiliationStyles();
  const affiliations = useSelector(getAffiliations);
  const dispatch = useDispatch();

  const editAffiliation = useCallback(
    (aff: Affiliation) => {
      dispatch(
        manuscriptEditorActions.showModalDialog({
          component: ConnectedAffiliationFormDialog,
          props: { affiliation: aff },
          title: 'Edit Affiliation'
        })
      );
    },
    [dispatch]
  );

  const addAffiliation = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ConnectedAffiliationFormDialog,
        props: {},
        title: 'Add Affiliation'
      })
    );
  }, [dispatch]);

  const renderAffiliation = (aff: Affiliation): React.ReactNode => (
    <li key={aff.id} className={classes.listItem}>
      <div className={classes.orderLabel}>{affiliations.length > 1 ? `(${aff.label})` : undefined}</div>
      <div className={classes.affiliationInfo}>{aff.getDisplayName()}</div>
      <IconButton classes={{ root: classes.editButton }} onClick={editAffiliation.bind(null, aff)}>
        <EditIcon fontSize="small" />
      </IconButton>
    </li>
  );

  return (
    <section>
      <SectionContainer label="Affiliations" id={id}>
        <ul className={classes.root}>{affiliations.map(renderAffiliation)}</ul>
      </SectionContainer>
      <ActionButton
        title="Add Affiliation"
        variant="addEntity"
        onClick={addAffiliation}
        className={classes.addButton}
      />
    </section>
  );
};
