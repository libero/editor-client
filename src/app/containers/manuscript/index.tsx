import React from 'react';
import { useSelector } from 'react-redux';
import { Backdrop, CircularProgress } from '@material-ui/core';

import { isManuscriptLoaded } from '../../selectors/manuscript.selectors';
import './styles.scss';
import { ManuscriptToolbar } from './manuscript-toolbar';
import { ManuscriptEditor } from './manuscript-editor';

const renderBackdrop = (): JSX.Element => (
  <Backdrop open={true}>
    <CircularProgress color="inherit" />
  </Backdrop>
);

export const ManuscriptContainer: React.FC = () => {
  const isLoaded = useSelector(isManuscriptLoaded);
  const renderContent = (): JSX.Element => (
    <React.Fragment>
      <ManuscriptToolbar />
      <ManuscriptEditor />
    </React.Fragment>
  );

  return isLoaded ? renderContent() : renderBackdrop();
};
