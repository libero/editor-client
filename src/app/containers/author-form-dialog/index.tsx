import React, { SyntheticEvent, useCallback, useState } from 'react';
import { TextField } from '@material-ui/core';
import { useDispatch, useSelector } from 'react-redux';

import { useAuthorFormStyles } from './styles';
import { createAuthor, Person } from 'app/models/person';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { PromptDialog } from 'app/components/prompt-dialog';
import { ActionButton } from 'app/components/action-button';
import { LinkedAffiliationsList } from './linked-affiliations-list';
import { getAffiliations, getAuthorAffiliations } from 'app/selectors/manuscript.selectors';
import { Affiliation } from 'app/models/affiliation';

interface AuthorFormDialogProps {
  author?: Person;
}

const renderConfirmDialog = (title: string, msg: string, onAccept: () => void, onReject: () => void) => {
  return (
    <PromptDialog
      title={title}
      message={msg}
      isOpen={true}
      onAccept={onAccept}
      onReject={onReject}
      acceptLabel="Delete"
      rejectLabel="Cancel"
      acceptVariant="containedWarning"
      rejectVariant="secondaryOutlined"
    />
  );
};

const labelProps = { shrink: true };

export const AuthorFormDialog: React.FC<AuthorFormDialogProps> = (props) => {
  const dispatch = useDispatch();
  const isNewAuthor = !props.author;

  const [author, setAuthor] = useState<Person>(
    props.author || createAuthor(undefined, { firstName: '', lastName: '' })
  );
  const [isConfirmShown, setConfirmSnow] = useState<boolean>(false);

  const allAffiliations = useSelector(getAffiliations);
  const linkedAffiliations = useSelector(getAuthorAffiliations)(author);
  const classes = useAuthorFormStyles();

  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleDelete = useCallback(() => {
    setConfirmSnow(true);
  }, [setConfirmSnow]);

  const handleAccept = useCallback(() => {
    setConfirmSnow(false);
    dispatch(manuscriptActions.deleteAuthorAction(author));
    closeDialog();
  }, [setConfirmSnow, author, closeDialog, dispatch]);

  const handleAffiliationsUpdate = useCallback(
    (affiliations: Affiliation[]) => {
      author.affiliations = affiliations.map(({ id }) => id);
    },
    [author]
  );

  const handleReject = useCallback(() => {
    setConfirmSnow(false);
  }, [setConfirmSnow]);

  const handleFormChange = useCallback(
    (event: SyntheticEvent) => {
      const fieldName = event.target['name'];
      const newValue = event.target['value'];
      setAuthor({
        ...author,
        [fieldName]: newValue
      });
    },
    [author, setAuthor]
  );

  const handleDone = useCallback(() => {
    if (isNewAuthor) {
      dispatch(manuscriptActions.addAuthorAction(author));
    } else {
      dispatch(manuscriptActions.updateAuthorAction(author));
    }
    closeDialog();
  }, [author, closeDialog, dispatch, isNewAuthor]);

  return (
    <section className={classes.root}>
      <TextField
        fullWidth
        name="firstName"
        label="Given name(s)"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={author.firstName}
        onChange={handleFormChange}
      />
      <TextField
        fullWidth
        name="lastName"
        label="Surname(s)"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={author.lastName}
        onChange={handleFormChange}
      />
      <TextField
        name="suffix"
        label="Suffix"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={author.suffix || ''}
        onChange={handleFormChange}
      />
      <TextField
        name="email"
        fullWidth
        label="Email"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={author.email || ''}
        onChange={handleFormChange}
      />
      <LinkedAffiliationsList
        allAffiliations={allAffiliations}
        linkedAffiliations={linkedAffiliations}
        onChange={handleAffiliationsUpdate}
      />
      <div className={classes.buttonPanel}>
        {!isNewAuthor ? <ActionButton variant="outlinedWarning" onClick={handleDelete} title="Delete" /> : undefined}
        <div aria-hidden={true} className={classes.spacer}></div>
        <ActionButton variant="secondaryOutlined" onClick={closeDialog} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleDone} title="Done" />
      </div>
      {isConfirmShown
        ? renderConfirmDialog(
            'You are deleting an author',
            'Deleting an author can leave unlinked affiliation and funding information. Are you sure you want to proceed?',
            handleAccept,
            handleReject
          )
        : undefined}
    </section>
  );
};
