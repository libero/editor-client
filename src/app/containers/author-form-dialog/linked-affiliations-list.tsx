import React, { useState, useCallback, useEffect } from 'react';
import MoreVertIcon from '@material-ui/icons/MoreVert';
import { get } from 'lodash';
import { FormControl, InputLabel, MenuItem, Select, IconButton, Menu } from '@material-ui/core';
import { useDispatch } from 'react-redux';

import { ActionButton } from 'app/components/action-button';
import { useAuthorFormStyles } from 'app/containers/author-form-dialog/styles';
import { Affiliation, getAffiliationDisplayName } from 'app/models/affiliation';
import { AffiliationFormDialog } from 'app/containers/affiliation-form-dialog/affiliation-form-dialog';
import { ModalContainer } from 'app/containers/modal-container';
import { ReactFCProps } from 'app/utils/types';
import * as manuscriptActions from 'app/actions/manuscript.actions';

interface LinkedAffiliationsListProps {
  linkedAffiliations: Affiliation[];
  allAffiliations: Affiliation[];
  onChange: (selectedAffiliations: Affiliation[]) => void;
}

const renderAffiliationModal = (props: ReactFCProps<typeof AffiliationFormDialog>) => {
  return (
    <ModalContainer
      title={props.affiliation ? 'Edit Affiliation' : 'Add Affiliation'}
      params={props}
      component={AffiliationFormDialog}
    />
  );
};

export const LinkedAffiliationsList: React.FC<LinkedAffiliationsListProps> = (props) => {
  const { linkedAffiliations, allAffiliations, onChange } = props;
  const classes = useAuthorFormStyles();
  const [userLinkedAffiliations, setUserLinkedAffiliations] = useState<Affiliation[]>([]);

  const [isAffiliationDialogShown, setAffiliationDialogShown] = useState<boolean>(false);
  const [editedAffiliation, setEditedAffiliation] = useState<Affiliation>(undefined);
  const [activeAffiliationIndex, setActiveAffiliationIndex] = useState<number>(-1);
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!userLinkedAffiliations.length) {
      setUserLinkedAffiliations(linkedAffiliations);
    }
  }, [linkedAffiliations, props, userLinkedAffiliations.length]);

  const triggerOnChange = useCallback(
    (affiliationsList) => {
      onChange(affiliationsList.filter((author) => Boolean(author)));
    },
    [onChange]
  );

  const handleAffiliationModalCancel = useCallback(() => {
    setAffiliationDialogShown(false);
  }, [setAffiliationDialogShown]);

  const handleAffiliationModalAccept = useCallback(
    (affiliation: Affiliation) => {
      if (!editedAffiliation) {
        dispatch(manuscriptActions.addAffiliationAction(affiliation));
      } else {
        dispatch(manuscriptActions.updateAffiliationAction(affiliation));
      }

      const updatedList = [...userLinkedAffiliations];
      updatedList[activeAffiliationIndex] = affiliation;
      setUserLinkedAffiliations(updatedList);
      triggerOnChange(updatedList);
      setAffiliationDialogShown(false);
    },
    [dispatch, userLinkedAffiliations, activeAffiliationIndex, triggerOnChange, editedAffiliation]
  );

  const getAuthorSelectList = useCallback(
    (selectedAuthor) => {
      const selectedAffiliationId = get(selectedAuthor, 'id');
      const linkedIds = new Set(userLinkedAffiliations.map((affiliation: Affiliation) => get(affiliation, 'id')));
      return allAffiliations.filter(({ id }) => !linkedIds.has(id) || selectedAffiliationId === id);
    },
    [userLinkedAffiliations, allAffiliations]
  );

  const handleAddNewAffiliation = useCallback(
    (index) => () => {
      setActiveAffiliationIndex(index);
      setEditedAffiliation(undefined);
      setAffiliationDialogShown(true);
    },
    [setActiveAffiliationIndex, setAffiliationDialogShown]
  );

  const renderAffiliationSelectListItem = useCallback(
    (affiliation: Affiliation) => {
      return (
        <MenuItem key={affiliation.id} value={affiliation.id} classes={{ root: classes.dropDownMenuItem }}>
          {getAffiliationDisplayName(affiliation)}
        </MenuItem>
      );
    },
    [classes]
  );

  const openMenu = useCallback(
    (index: number) => (event: React.MouseEvent<HTMLButtonElement>) => {
      setActiveAffiliationIndex(index);
      setAnchorEl(event.currentTarget);
    },
    []
  );

  const deleteRow = useCallback(() => {
    const updatedList = [...userLinkedAffiliations];
    updatedList.splice(activeAffiliationIndex, 1);
    setUserLinkedAffiliations(updatedList);
    triggerOnChange(updatedList);
    setAnchorEl(null);
  }, [userLinkedAffiliations, activeAffiliationIndex, triggerOnChange, setAnchorEl]);

  const handleCloseMenu = useCallback(() => {
    setAnchorEl(null);
  }, [setAnchorEl]);

  const handleDeleteAffiliation = useCallback(
    (affiliation: Affiliation) => {
      const updatedList = [...userLinkedAffiliations];
      updatedList[activeAffiliationIndex] = null;
      setUserLinkedAffiliations(updatedList);
      setActiveAffiliationIndex(-1);
      setAffiliationDialogShown(false);
      dispatch(manuscriptActions.deleteAffiliationAction(affiliation));
    },
    [userLinkedAffiliations, activeAffiliationIndex, dispatch]
  );

  const updateAffiliation = useCallback(() => {
    setAnchorEl(null);
    setEditedAffiliation(userLinkedAffiliations[activeAffiliationIndex]);
    setAffiliationDialogShown(true);
  }, [setEditedAffiliation, userLinkedAffiliations, activeAffiliationIndex]);

  const updateRow = useCallback(
    (index) => (event) => {
      const updatedList = [...userLinkedAffiliations];
      updatedList[index] = allAffiliations.find(({ id }) => id === event.target.value);
      setUserLinkedAffiliations(updatedList);
      triggerOnChange(updatedList);
    },
    [userLinkedAffiliations, allAffiliations, triggerOnChange]
  );

  const addEmptyRow = useCallback(() => {
    setUserLinkedAffiliations([...userLinkedAffiliations, null]);
  }, [userLinkedAffiliations, setUserLinkedAffiliations]);

  return (
    <div className={classes.inputField}>
      <Menu id="simple-menu" anchorEl={anchorEl} keepMounted open={Boolean(anchorEl)} onClose={handleCloseMenu}>
        <MenuItem onClick={updateAffiliation} disabled={!userLinkedAffiliations[activeAffiliationIndex]}>
          Edit Affiliation
        </MenuItem>
        <MenuItem onClick={deleteRow}>Remove Affiliation</MenuItem>
      </Menu>
      {userLinkedAffiliations.map((affiliation: Affiliation, index: number) => (
        <div className={classes.affiliatedAuthorRow} key={affiliation?.id || index}>
          <FormControl key={get(affiliation, 'id', index)} variant="outlined" className={classes.affiliatedAuthorInput}>
            <InputLabel shrink id="author-affiliations-label">
              Affiliation
            </InputLabel>
            <Select
              labelId="author-affiliations-label"
              displayEmpty
              value={get(affiliation, 'id', '')}
              onChange={updateRow(index)}
              label="Affiliation"
            >
              <MenuItem value={''}>Set affiliation</MenuItem>
              <MenuItem onMouseDown={handleAddNewAffiliation(index)}>Add new affiliation</MenuItem>
              {getAuthorSelectList(affiliation).map(renderAffiliationSelectListItem)}
            </Select>
          </FormControl>
          <IconButton classes={{ root: classes.deleteButton }} onClick={openMenu(index)}>
            <MoreVertIcon fontSize="small" />
          </IconButton>
        </div>
      ))}
      <ActionButton variant="addEntity" title="Affiliation" onClick={addEmptyRow} />
      {isAffiliationDialogShown
        ? renderAffiliationModal({
            affiliation: editedAffiliation,
            onAccept: handleAffiliationModalAccept,
            onCancel: handleAffiliationModalCancel,
            onDelete: handleDeleteAffiliation
          })
        : undefined}
    </div>
  );
};
