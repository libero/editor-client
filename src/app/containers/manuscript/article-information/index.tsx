import React, { useCallback } from 'react';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';
import { useSelector, useDispatch } from 'react-redux';

import { SectionContainer } from 'app/components/section-container';
import { getArticleInformation } from 'app/selectors/manuscript.selectors';
import { useArticleInformationStyles } from 'app/containers/manuscript/article-information/styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ArticleInfoFormDialog } from 'app/containers/article-info-form-dialog';

export const ArticleInformation: React.FC<{}> = () => {
  const articleInfo = useSelector(getArticleInformation);
  const dispatch = useDispatch();
  const editArticleInfo = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ArticleInfoFormDialog,
        title: 'Article Information'
      })
    );
  }, [dispatch]);

  const classes = useArticleInformationStyles();
  return (
    <SectionContainer label="Article information" className={classes.sectionContainer}>
      <section className={classes.infoSection}>
        <div>
          <strong> Article DOI: </strong> {articleInfo.articleDOI}
        </div>
        <div>
          <strong> Publisher ID: </strong> {articleInfo.publisherId}
        </div>
      </section>
      <IconButton classes={{ root: classes.editButton }} onClick={editArticleInfo}>
        <EditIcon fontSize="small" />
      </IconButton>
    </SectionContainer>
  );
};
