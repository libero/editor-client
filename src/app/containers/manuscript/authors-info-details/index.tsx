import React, { useCallback } from 'react';
import { IconButton } from '@material-ui/core';
import { useSelector, useDispatch } from 'react-redux';
import EditIcon from '@material-ui/icons/Edit';
import { isBoolean, isEqual } from 'lodash';
import Interweave from 'interweave';

import { SectionContainer } from 'app/components/section-container';
import { getAuthorAffiliations, getAuthors } from 'app/selectors/manuscript.selectors';
import { Person } from 'app/models/person';
import {
  useAuthorDetailStyles,
  useAuthorsDetailsListStyles
} from 'app/containers/manuscript/authors-info-details/styles';
import { ActionButton } from 'app/components/action-button';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { AuthorFormDialog } from 'app/containers/author-form-dialog';
import { stringifyEditorState } from 'app/utils/view.utils';
import { OrcidIcon } from 'app/assets/icons';
import { ComponentWithId } from 'app/types/utility.types';

export const AuthorsInfoDetails: React.FC<ComponentWithId> = ({ id }) => {
  const authors = useSelector(getAuthors);
  const classes = useAuthorsDetailsListStyles();
  const dispatch = useDispatch();

  const addAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        title: 'Add Author'
      })
    );
  }, [dispatch]);

  return (
    <section>
      <SectionContainer label="Author Information" id={id}>
        {authors.map((author) => (
          <AuthorInfo key={author.id} author={author} />
        ))}
      </SectionContainer>
      <ActionButton variant="addEntity" title="Add Author" onClick={addAuthor} className={classes.addAuthorButton} />
    </section>
  );
};

interface AuthorInfoProps {
  author: Person;
}

const AuthorInfo: React.FC<AuthorInfoProps> = React.memo(({ author }) => {
  const classes = useAuthorDetailStyles();
  const dispatch = useDispatch();
  const affiliations = useSelector(getAuthorAffiliations)(author);

  const editAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        props: { author },
        title: 'Edit Author'
      })
    );
  }, [author, dispatch]);

  return (
    <section className={classes.root}>
      <div className={classes.authorInfoContainer}>
        <div className={classes.authorInfoLine}>
          <strong> {author.getDisplayName()} </strong>
        </div>
        <div className={classes.authorInfoLine}>
          {author.isCorrespondingAuthor ? 'Corresponding Author: ' : undefined}
          {author.email}
        </div>
        <div className={classes.authorInfoLine}>
          {affiliations.map((aff) => (
            <div key={aff.id}>{aff.getDisplayName()}</div>
          ))}
        </div>
        <div className={classes.authorInfoLine}>
          {author.isAuthenticated ? <OrcidIcon className={classes.orcidIcon} /> : undefined}{' '}
          <a href={`https://orcid.org/${author.orcid}`} target="_blank" rel="noopener noreferrer">
            {author.orcid}
          </a>
        </div>
        {isBoolean(author.hasCompetingInterest) ? (
          <div className={classes.authorInfoLine}>
            {author.hasCompetingInterest
              ? `Competing interest: ${author.competingInterestStatement}`
              : 'No Competing interest declared'}
          </div>
        ) : undefined}
        {author.bio ? (
          <div className={classes.authorBio}>
            <Interweave content={stringifyEditorState(author.bio)} />
          </div>
        ) : undefined}
      </div>
      <IconButton classes={{ root: classes.editButton }} onClick={editAuthor}>
        <EditIcon fontSize="small" />
      </IconButton>
    </section>
  );
}, isEqual);
