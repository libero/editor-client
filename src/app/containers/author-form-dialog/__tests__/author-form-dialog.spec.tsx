import React from 'react';
import { create } from 'react-test-renderer';
import { EditorState } from 'prosemirror-state';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import { AuthorFormDialog } from 'app/containers/author-form-dialog/index';
import { addAuthorAction, deleteAuthorAction, updateAuthorAction } from 'app/actions/manuscript.actions';
import { PromptDialog } from 'app/components/prompt-dialog';
import { givenState } from 'app/test-utils/reducer-test-helpers';
import { ManuscriptHistoryState } from 'app/store';

jest.mock('app/components/rich-text-input', () => ({
  RichTextInput: () => <div data-cmp="rich-text-input"></div>
}));

jest.mock('@material-ui/core', () => ({
  Select: ({ children }) => <div data-cmp="Index">{children}</div>,
  Dialog: () => <div data-cmp="Dialog"></div>,
  Menu: ({ children }) => <div data-cmp="Menu">{children}</div>,
  MenuItem: ({ children }) => <div data-cmp="MenuItem">{children}</div>,
  Checkbox: ({ children }) => <div data-cmp="Checkbox" />,
  FormControlLabel: ({ children }) => <div data-cmp="FormControlLabel" />,
  FormControl: ({ children }) => <div data-cmp="FormControl">{children}</div>,
  InputLabel: ({ children }) => <div data-cmp="InputLabel">{children}</div>,
  TextField: () => <div data-cmp="TextField"></div>,
  IconButton: () => <div data-cmp="iconButton"></div>,
  Button: () => <div data-cmp="Button"></div>
}));

describe('Author Form Dialog', () => {
  const mockStore = configureMockStore([]);
  let mockState: ManuscriptHistoryState;

  beforeEach(() => {
    mockState = givenState({
      authors: [
        {
          id: '4d53e405-5225-4858-a87a-aec902ae50b6',
          firstName: 'Fred',
          lastName: 'Atherden',
          email: 'f.atherden@elifesciences.org',
          orcid: 'https://orcid.org/0000-0002-6048-1470'
        }
      ]
    });
  });

  it('renders new author dialog form', () => {
    const store = mockStore({
      manuscript: mockState
    });

    const wrapper = create(
      <Provider store={store}>
        <AuthorFormDialog />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a form with competing interest statement', () => {
    mockState.data.present.authors[0].hasCompetingInterest = true;
    mockState.data.present.authors[0].competingInterestStatement = 'statement';

    const store = mockStore({
      manuscript: mockState
    });

    const wrapper = create(
      <Provider store={store}>
        <AuthorFormDialog />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders edit author dialog form', () => {
    const store = mockStore({
      manuscript: mockState
    });

    const wrapper = create(
      <Provider store={store}>
        <AuthorFormDialog author={mockState.data.present.authors[0]} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('dispatches an event to create new author', () => {
    const store = mockStore({
      manuscript: mockState
    });
    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <AuthorFormDialog />
      </Provider>
    );
    wrapper.find({ title: 'Done' }).prop('onClick')();
    expect(store.dispatch).toBeCalledWith(
      addAuthorAction({
        firstName: '',
        id: expect.any(String),
        lastName: '',
        affiliations: [],
        bio: expect.any(EditorState)
      })
    );
  });

  it('dispatches an event to save edited author', () => {
    const store = mockStore({
      manuscript: mockState
    });
    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <AuthorFormDialog author={mockState.data.present.authors[0]} />
      </Provider>
    );
    wrapper.find({ title: 'Done' }).prop('onClick')();
    expect(store.dispatch).toBeCalledWith(updateAuthorAction(mockState.data.present.authors[0]));
  });

  it('dispatches an event to delete author', () => {
    const store = mockStore({
      manuscript: mockState
    });
    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <AuthorFormDialog author={mockState.data.present.authors[0]} />
      </Provider>
    );

    wrapper.find({ title: 'Delete' }).prop('onClick')();
    wrapper.update();

    wrapper.find(PromptDialog).prop('onAccept')();
    expect(store.dispatch).toBeCalledWith(deleteAuthorAction(mockState.data.present.authors[0]));
  });
});
