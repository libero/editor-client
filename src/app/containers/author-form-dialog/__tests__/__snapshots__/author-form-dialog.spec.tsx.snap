// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Author Form Dialog renders edit author dialog form 1`] = `
<section
  className="makeStyles-root-1"
>
  <div
    data-cmp="TextField"
  />
  <div
    data-cmp="TextField"
  />
  <div
    data-cmp="TextField"
  />
  <div
    data-cmp="TextField"
  />
  <div
    data-cmp="rich-text-input"
  />
  <div
    className="makeStyles-inputField-2"
  >
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Edit Affiliation
      </div>
      <div
        data-cmp="MenuItem"
      >
        Remove Affiliation
      </div>
    </div>
    <div
      className="makeStyles-affiliatedAuthorRow-5"
    >
      <div
        data-cmp="FormControl"
      >
        <div
          data-cmp="InputLabel"
        >
          Affiliation
        </div>
        <div
          data-cmp="Select"
        >
          <div
            data-cmp="MenuItem"
          >
            Set affiliation
          </div>
          <div
            data-cmp="MenuItem"
          >
            Add new affiliation
          </div>
        </div>
      </div>
      <div
        data-cmp="iconButton"
      />
    </div>
    <div
      data-cmp="Button"
    />
  </div>
  <div
    className="makeStyles-buttonPanel-3"
  >
    <div
      data-cmp="Button"
    />
    <div
      aria-hidden={true}
      className="makeStyles-spacer-4"
    />
    <div
      data-cmp="Button"
    />
    <div
      data-cmp="Button"
    />
  </div>
</section>
`;

exports[`Author Form Dialog renders new author dialog form 1`] = `
<section
  className="makeStyles-root-1"
>
  <div
    data-cmp="TextField"
  />
  <div
    data-cmp="TextField"
  />
  <div
    data-cmp="TextField"
  />
  <div
    data-cmp="TextField"
  />
  <div
    data-cmp="rich-text-input"
  />
  <div
    className="makeStyles-inputField-2"
  >
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Edit Affiliation
      </div>
      <div
        data-cmp="MenuItem"
      >
        Remove Affiliation
      </div>
    </div>
    <div
      data-cmp="Button"
    />
  </div>
  <div
    className="makeStyles-buttonPanel-3"
  >
    <div
      aria-hidden={true}
      className="makeStyles-spacer-4"
    />
    <div
      data-cmp="Button"
    />
    <div
      data-cmp="Button"
    />
  </div>
</section>
`;
